store/postgres/src/dynds.rs

```diff
@@ -0,0 +1,131 @@
+//! SQL queries to load dynamic data sources
+
+use diesel::pg::PgConnection;
+use diesel::prelude::{ExpressionMethods, JoinOnDsl, QueryDsl, RunQueryDsl};
+
+use graph::{
+    components::store::StoredDynamicDataSource,
+    data::subgraph::Source,
+    prelude::{bigdecimal::ToPrimitive, web3::types::H160, BigDecimal, StoreError},
+};
+
+// Diesel tables for some of the metadata
+// See also: ed42d219c6704a4aab57ce1ea66698e7
+// Changes to the GraphQL schema might require changes to these tables.
+// The definitions of the tables can be generated with
+//    cargo run -p graph-store-postgres --example layout -- \
+//      -g diesel store/postgres/src/subgraphs.graphql subgraphs
+// BEGIN GENERATED CODE
+table! {
+    subgraphs.dynamic_ethereum_contract_data_source (vid) {
+        vid -> BigInt,
+        id -> Text,
+        kind -> Text,
+        name -> Text,
+        network -> Nullable<Text>,
+        source -> Text,
+        mapping -> Text,
+        templates -> Nullable<Array<Text>>,
+        ethereum_block_hash -> Binary,
+        ethereum_block_number -> Numeric,
+        deployment -> Text,
+        context -> Nullable<Text>,
+        block_range -> Range<Integer>,
+    }
+}
+
+table! {
+    subgraphs.ethereum_contract_source (vid) {
+        vid -> BigInt,
+        id -> Text,
+        address -> Nullable<Binary>,
+        abi -> Text,
+        start_block -> Nullable<Numeric>,
+        block_range -> Range<Integer>,
+    }
+}
+
+// END GENERATED CODE
+
+allow_tables_to_appear_in_same_query!(
+    dynamic_ethereum_contract_data_source,
+    ethereum_contract_source
+);
+
+fn to_source(
+    deployment: &str,
+    ds_id: &str,
+    (address, abi, start_block): (Option<Vec<u8>>, String, Option<BigDecimal>),
+) -> Result<Source, StoreError> {
+    // Treat a missing address as an error. TODO: Is that correct?
```

Comment:
This is correct for dynamic data sources.

Reply:
Updated comment